src/components/FeatureViews/Courses/AttendanceTable.js

```diff
@@ -0,0 +1,437 @@
+  query getAttendance($courseId: ID!) {
+    __typename
+    attendances(courseId: $courseId) {
+      id
+      status
+      session {
+        startDatetime
+        id
+      }
+      enrollment {
+        student {
+          user {
+            lastName
+            id
+            firstName
+          }
+        }
+      }
+    }
+    sessions(courseId: $courseId) {
+      id
+      startDatetime
+    }
+    enrollments(courseId: $courseId) {
+      student {
+        user {
+          firstName
+          id
+          lastName
+        }
+      }
+    }
+  }
+`;
+
+export const UPDATE_ATTENDANCE_STATUS = gql`
+  mutation UpdateAttendanceStatus(
+    $attendanceId: ID!
+    $status: AttendanceStatusEnum!
+  ) {
+    __typename
+    updateAttendance(id: $attendanceId, status: $status) {
+      attendance {
+        updatedAt
+      }
+    }
+  }
+`;
+
+const AttendanceTable = ({ setIsEditing }) => {
+  const { id } = useParams();
+  const classes = useStyles();
+  const [attendanceEditStates, setAttendanceEditStates] = useState();
+  const [isCheckBoxEditing, setIsCheckBoxEditing] = useState();
+  const [courseAttendanceMatrix, setCourseAttendanceMatrix] = useState([]);
+  const [sortByAlphabet, setSortByAlphabet] = useState("");
+  const [attendanceBySession, setAttendanceBySession] = useState({});
+
+  useEffect(() => {
+    if(courseAttendanceMatrix.length > 0) {
+      let sessionsDoneEditing = {};
+      courseAttendanceMatrix.forEach((studentAttendanceRow) => {
+        const { attendanceList: studentAttendanceList } = studentAttendanceRow;
+        studentAttendanceList.forEach((attendanceStatus) => {
+          const { sessionId } = attendanceStatus;
+          if (attendanceStatus[sessionId] !== "UNSET") {
+            sessionsDoneEditing[sessionId] = "done";
+          };
+        });
+      });
+      setAttendanceEditStates({ ...attendanceEditStates, ...sessionsDoneEditing });
+    }
+  }, [courseAttendanceMatrix]);
+
+  const sortByFirstName = (firstStudent, secondStudent) =>
+  { 
+    if(firstStudent.studentName < secondStudent.studentName) {
+      return -1;
+    } else if(firstStudent.studentName > secondStudent.studentName) {
+      return 1;
+    } else {
+      return 0;
+    };
+  };
+
+  const { data, loading, error } = useQuery(GET_ATTENDANCE, {
+    variables: { courseId: id },
+    onCompleted: () => {
+      const { sessions, enrollments, attendances } = data;
+      setAttendanceEditStates(
+        sessions.reduce(
+          (accum, currentValue) => ({
+            ...accum,
+            [currentValue.id]: "noSelect",
+          }),
+          {}
+        )
+      );
+
+      setIsCheckBoxEditing(
+        sessions.reduce(
+          (allCheckboxStatus, {id: sessionId}) => ({ ...allCheckboxStatus, [sessionId]: false }),{})
+      );
+
+      const studentAttendanceRowsData = attendances.reduce(
+        (allStudentAttendanceRowsData, { id, enrollment, status, session }) => {
```

Comment:
similar story - pull this function in your .reduce() into an explanatory variable. You're making the reader read through all of this to understand how studentAttendanceRowsData is set

Reply:
also don't destructure the current attendance object. "id" "enrollment" "status" "session" are all vague so seeing "attendance.id" helps the reader know what it's the id of